helix-term/src/commands.rs

```diff
@@ -519,6 +521,33 @@ fn goto_line_start(cx: &mut Context) {
     )
 }
 
+fn goto_next_buffer(cx: &mut Context) {
+    goto_buffer(cx, Direction::Forward);
+}
+
+fn goto_previous_buffer(cx: &mut Context) {
+    goto_buffer(cx, Direction::Backward);
+}
+
+fn goto_buffer(cx: &mut Context, direction: Direction) {
+    let buf_cur = current!(cx.editor).1.id();
+
+    if let Some(pos) = cx.editor.documents.iter().position(|(id, _)| id == buf_cur) {
+        let goto_id = if direction == Direction::Forward {
+            if pos < cx.editor.documents.iter().count() - 1 {
+                cx.editor.documents.iter().nth(pos + 1).unwrap().0
+            } else {
+                cx.editor.documents.iter().next().unwrap().0
+            }
+        } else if pos > 0 {
+            cx.editor.documents.iter().nth(pos - 1).unwrap().0
+        } else {
+            cx.editor.documents.iter().last().unwrap().0
+        };
```

Comment:
This can be further simplified and optimized to only iterate once.

```rust
let current = view!(cx.editor).doc; // a lot cheaper than calling `current!().1.id()`

let id = match direction {
    Direction::Forward => {
        let mut iter = cx.editor.documents.keys();
        iter.skip_while(|id| id != current);
        iter.next();
        iter.next().or(cx.editor.documents.keys().first())
    }
    Direction::Backward => {
        // slot_map doesn't implement DoubleEndedIterator so no .rev()
        let mut iter = cx.editor.documents.keys().peekable();
        // keep iterating until the next item is the current id
        while Some(id) = iter.peek() {
            if id == current {
              break;
            }
            iter.next();
        }
        iter.next()
    }
}.unwrap();

cx.editor.switch(goto_id, Action::Replace);
 ```

Reply:
i tried your suggestion and came to this:

    let current = view!(cx.editor).doc;

    let id = if let Direction::Forward = direction {
        let mut iter = cx.editor.documents.keys();
        iter.skip_while(|id| id != &current);
        iter.next();
        iter.next().or(cx.editor.documents.keys().next())
    } else {
        let mut iter = cx.editor.documents.keys().peekable();
        while let Some(id) = iter.peek() {
            if id == &current {
                break;
            }
            iter.next();
        }
        iter.next()
    }
    .unwrap();

    cx.editor.switch(id, Action::Replace);

Howerver, i dont know how to fix the "borrow of moved value: iter" on line with "iter.next();".

Reply:
Also, if we peek the next buffer, and thats the buffer we are currently on, it breaks out from the while loop, but still iterates to the next buffer after it.
And if we are on the first buffer and try to change to previous buffer then it wont wrap us back to the last buffer i think.

Reply:
Yeah, it needs a clause before the while loop:

```rust
if iter.peek() == Some(current) {
  iter.last()
} else {
  // rest of the code
}
```

Reply:
yes, but that still only takes us the same buffer we are currently on :(

Reply:
No, it will iterate until the last element.

Reply:
Yes, i mean if we are not on the first buffer, then it wont change the buffer.

Reply:
If it peeks and we are on the first buffer, then `iter.last()` is called, returning the last buffer on the list.

Reply:
I replaced SlotMap with BTreeMap which implements DoubleEndedIterator so we can use a similar implementation on both ends https://github.com/helix-editor/helix/commit/e2560f427ef5e75155071e39da342628f5d5896a

Reply:
```rust
    let current = view!(cx.editor).doc;

    let id = if let Direction::Forward = direction {
        let mut iter = cx.editor.documents.keys();
        iter.skip_while(|id| id != &current);
        iter.next();
        iter.next().or(cx.editor.documents.keys().next())
    } else {
        let mut iter = cx.editor.documents.keys().rev();
        iter.skip_while(|id| id != &current);
        iter.next();
        iter.next().or(cx.editor.documents.keys().rev().next())
    }.unwrap();

    cx.editor.switch(id, Action::Replace);
```

Reply:
> Howerver, i dont know how to fix the "borrow of moved value: iter" on line with "iter.next(); inside the Direction::Forward part of the code".

Any idea about how to fix this issue?

Reply:
Pushed